runtime/compiler/optimizer/HotFieldMarking.cpp

```diff
@@ -0,0 +1,173 @@
+   {
+   return (currentValue + newFrequency);
+   }
+
+static int32_t getReducedFrequencyAverage(int32_t currentValue, int32_t count, int32_t newFrequency)
+   {
+   return ((currentValue * count) + newFrequency) / (count + 1);
+   }
+
+int32_t TR_HotFieldMarking::perform()
+   {
+   if (!jitConfig->javaVM->memoryManagerFunctions->j9gc_hot_reference_field_required(jitConfig->javaVM))
+      {
+      if (trace())
+         traceMsg(comp(), "Skipping hot field marking since dynamic breadth first scan ordering is disabled\n");
+      return 0;
+      }
+
+   SymAggMap stats(SymAggMapComparator(), comp()->trMemory()->currentStackRegion());
+   TR::Block *block = NULL;
+   for (TR::PostorderNodeIterator it(comp()->getStartTree(), comp()); it != NULL; ++it)
+      {
+      TR::Node * const node = it.currentNode();
+      if (node->getOpCodeValue() == TR::BBStart)
+         {
+         block = node->getBlock();
+         }
+      else if ((node->getOpCode().isLoadIndirect() || node->getOpCode().isStoreIndirect())
+               && node->getOpCode().hasSymbolReference()
+               && !node->getSymbolReference()->isUnresolved()
+               && node->getSymbolReference()->getSymbol()->isShadow()
+               && !node->isInternalPointer()
+               && !node->getOpCode().isArrayLength()
+               && node->getSymbolReference()->getSymbol()->isCollectedReference()
+              )
+         {
+         TR::SymbolReference *symRef = node->getSymbolReference();
+         if (symRef->getCPIndex() >= 0 && !symRef->getSymbol()->isArrayShadowSymbol())
+            {
+            auto itr = stats.find(symRef->getSymbol());
+            if (itr != stats.end())
+               {
+               if(TR::Options::getReductionAlgorithm(TR_HotFieldReductionAlgorithmSum))
+                  {
+                  itr->second->_score = getReducedFrequencySum(itr->second->_score, block->getFrequency());
+                  } 
+               else
+                  {
+                  itr->second->_score = getReducedFrequencyAverage(itr->second->_score, itr->second->_count, block->getFrequency());
+                  itr->second->_count += 1;
+                  }
+               continue;
+               }
+
+            TR::ResolvedMethodSymbol *rms = comp()->getOwningMethodSymbol(symRef->getOwningMethodIndex());
+            TR_ResolvedMethod *method = rms->getResolvedMethod();
+
+            int32_t fieldNameLength = 0;
+            char *fieldName = method->fieldNameChars(symRef->getCPIndex(), fieldNameLength);
+            int32_t fieldSigLength = 0;
+            char *fieldSig = method->fieldSignatureChars(symRef->getCPIndex(), fieldSigLength);
+            bool isStatic = false;
+            TR_OpaqueClassBlock *containingClass = static_cast<TR_ResolvedJ9Method*>(method)->definingClassFromCPFieldRef(comp(), symRef->getCPIndex(), isStatic);
+            if (isStatic)
+               continue;
```

Comment:
this code needs more tracing about the fields being processed and the values of the stats after each field is processed to facilitate debugging in the future.